File: pronacej-front/src/app/modules/general/services/informe.service.ts
import { Injectable } from '@angular/core';
import { BodyEncriptado } from 'app/core/model/both/bodyEncriptado.model';
import { DocumentoDTO } from 'app/core/model/both/DocumentoDTO.model';
import { CampoInformeDTO } from 'app/core/model/both/informe/campoInformeDTO.model';
import { InformeDTO } from 'app/core/model/both/informe/informeDTO.model';
import { PlantillaInformeDTO } from 'app/core/model/both/informe/plantillaInformeDTO.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable, Subscriber } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class InformeService {
    private path = '/informe';

    constructor(private backendService: BackendService) {}

    obtenerInformes(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<InformeDTO>>> {
        let endPoint = this.path + '/obtenerInformes';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerInformesPorToken(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<InformeDTO>>> {
        let endPoint = this.path + '/obtenerInformesPorToken';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    crearInforme(
        informeDTO: InformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<InformeDTO>> {
        let endPoint = this.path + '/crearInforme';
        return this.backendService.postFinal(
            endPoint,
            informeDTO,
            nemonicoMenu
        );
    }

    crearInformePorToken(
        informeDTO: InformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<InformeDTO>> {
        let endPoint = this.path + '/crearInformePorToken';
        return this.backendService.postFinal(
            endPoint,
            informeDTO,
            nemonicoMenu
        );
    }

    actualizarInforme(
        informeDTO: InformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<InformeDTO>> {
        let endPoint = this.path + '/actualizarInforme';
        return this.backendService.postFinal(
            endPoint,
            informeDTO,
            nemonicoMenu
        );
    }

    /**
     * Sube un informe firmado
     *
     * @param informeDTO InformeDTO
     * @param archivo File
     * @param nemonicoMenu string nemonico menu
     *
     * @return Observable<RespuestaPorDefecto<Boolean>>
     */
    subirInformeFirmado(
        informeDTO: InformeDTO,
        archivo: File,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/subirInformeFirmado';
        return new Observable(
            (subs: Subscriber<RespuestaPorDefecto<Boolean>>) => {
                let formData = new FormData();
                formData.append('documento', archivo);
                this.backendService
                    .crearBodyEncriptado(informeDTO)
                    .then((bodyEncriptado) => {
                        formData.append('body', JSON.stringify(bodyEncriptado));
                        this.backendService
                            .postFormDataBodyEncriptado2(
                                endPoint,
                                formData,
                                nemonicoMenu
                            )
                            .subscribe({
                                next: async (body: BodyEncriptado) => {
                                    let resp =
                                        await this.backendService.desencriptarBdyEncriptado<
                                            RespuestaPorDefecto<Boolean>
                                        >(body);
                                    subs.next(resp);
                                    subs.complete();
                                },
                                error: (error: any) => {
                                    subs.error(error);
                                    subs.complete();
                                },
                            });
                    })
                    .catch((error: any) => {
                        subs.error(error);
                        subs.complete();
                    });
            }
        );
    }

    obtenerDocumentos(
        request: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PaginacionResponse<DocumentoDTO>>> {
        let endPoint = this.path + '/obtenerDocumentos';
        return this.backendService.postFinal(endPoint, request, nemonicoMenu);
    }

    eliminarInforme(
        informeListaDTO: InformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/eliminarInforme';
        return this.backendService.postFinal(
            endPoint,
            informeListaDTO,
            nemonicoMenu
        );
    }

    obtenerListaPlantillas(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<
        RespuestaPorDefecto<PaginacionResponse<PlantillaInformeDTO>>
    > {
        let endPoint = this.path + '/obtenerListaPlantillas';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    obtenerPlantillas(
        nemonicoMenu: string,
        tokenCentro: string = null
    ): Observable<RespuestaPorDefecto<PlantillaInformeDTO[]>> {
        let endPoint = this.path + '/obtenerPlantillas';

        // Agregar tokenCentro solo si tiene un valor
        if (tokenCentro) {
            endPoint += `?tokenCentro=${tokenCentro}`;
        }

        return this.backendService.getFinal(endPoint, {}, nemonicoMenu);
    }

    crearPlantilla(
        plantillaInformeDTO: PlantillaInformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<PlantillaInformeDTO>> {
        let endPoint = this.path + '/crearPlantilla';
        return this.backendService.postFinal(
            endPoint,
            plantillaInformeDTO,
            nemonicoMenu
        );
    }

    actualizarPlantilla(
        plantillaInformeDTO: PlantillaInformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/actualizarPlantilla';
        return this.backendService.postFinal(
            endPoint,
            plantillaInformeDTO,
            nemonicoMenu
        );
    }

    eliminarPlantilla(
        plantillaInformeDTO: PlantillaInformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/eliminarPlantilla';
        return this.backendService.postFinal(
            endPoint,
            plantillaInformeDTO,
            nemonicoMenu
        );
    }

    obtenerCamposPorPlantilla(
        plantillaInformeDTO: PlantillaInformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<CampoInformeDTO[]>> {
        let endPoint = this.path + '/obtenerCamposPorIdPlantilla';
        return this.backendService.postFinal(
            endPoint,
            plantillaInformeDTO,
            nemonicoMenu
        );
    }

    obtenerCamposPorInforme(
        informeDTO: InformeDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<CampoInformeDTO[]>> {
        let endPoint = this.path + '/obtenerCamposPorIdInforme';
        return this.backendService.postFinal(
            endPoint,
            informeDTO,
            nemonicoMenu
        );
    }

    obtenerCamposPorNemonico(
        nemonicoInforme: string,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<CampoInformeDTO[]>> {
        let endPoint = this.path + '/obtenerCamposPorNemonico';
        return this.backendService.postFinal(
            endPoint,
            nemonicoInforme,
            nemonicoMenu
        );
    }
}

File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/general/InformeService.java
package net.latinus.sistema.integral.gestion.seguridad.service.general;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.DocumentoDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.both.informe.CampoInformeDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.both.informe.InformeDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.both.informe.PlantillaInformeDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface InformeService {

    //region Informe

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<InformeDTO>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<InformeDTO>> obtenerInformes(HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<InformeDTO>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<InformeDTO>> obtenerInformesPorToken(HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<InformeDTO>
     */
    RespuestaPorDefectoAuditoria<InformeDTO> obtenerInformePorId(HttpServletRequest httpServletRequest,
                                                                 InformeDTO informeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeDTO si la encuesta se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param informeDTO         objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<InformeDTO>
     */
    RespuestaPorDefectoAuditoria<InformeDTO> crearInforme(HttpServletRequest httpServletRequest,
                                                          InformeDTO informeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeDTO si la encuesta se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param informeDTO         objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<InformeDTO>
     */
    RespuestaPorDefectoAuditoria<InformeDTO> crearInformePorToken(HttpServletRequest httpServletRequest,
                                                                  InformeDTO informeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data InformeDTO si la empresa se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param informeDTO         objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<InformeDTO>
     */
    RespuestaPorDefectoAuditoria<InformeDTO> actualizarInforme(HttpServletRequest httpServletRequest,
                                                               InformeDTO informeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si el documento se sube con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado         objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> subirInformeFirmado(HttpServletRequest httpServletRequest,
                                                              MultipartFile multipartFile,
                                                              BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si el documento se sube con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado         objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<DocumentoDTO>> obtenerDocumentos(HttpServletRequest httpServletRequest,
                                                                                     BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si el informe se remueve con exito
     *
     * @param httpServletRequest request peticion.
     * @param informeDTO         objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> removerInforme(HttpServletRequest httpServletRequest,
                                                         InformeDTO informeDTO);

    //endregion

    //region PlantillaInforme

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data PlantillaInformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<PlantillaInformeDTO>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<PlantillaInformeDTO>> obtenerListaPlantillasInforme(HttpServletRequest httpServletRequest, BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data PlantillaInformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<PlantillaInformeDTO>
     */
    RespuestaPorDefectoAuditoria<List<PlantillaInformeDTO>> obtenerPlantillasInforme(HttpServletRequest httpServletRequest, String tokenCentro);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data PlantillaInformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<PlantillaInformeDTO>
     */
    RespuestaPorDefectoAuditoria<PlantillaInformeDTO> obtenerPlantillaInformePorId(HttpServletRequest httpServletRequest,
                                                                                   PlantillaInformeDTO plantillaInformeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data PlantillaInformeDTO si la encuesta se creo con exito
     *
     * @param httpServletRequest  request peticion.
     * @param plantillaInformeDTO objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<PlantillaInformeDTO>
     */
    RespuestaPorDefectoAuditoria<PlantillaInformeDTO> crearPlantillaInforme(HttpServletRequest httpServletRequest,
                                                                            PlantillaInformeDTO plantillaInformeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si la plantilla se creo con exito
     *
     * @param httpServletRequest  request peticion.
     * @param bodyEncriptado objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> actualizarPlantillaInforme(HttpServletRequest httpServletRequest,
                                                                     BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si la plantilla se remueve con exito
     *
     * @param httpServletRequest  request peticion.
     * @param plantillaInformeDTO objeto encuesta dto.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> removerPlantillaInforme(HttpServletRequest httpServletRequest,
                                                                  PlantillaInformeDTO plantillaInformeDTO);
    //endregion

    //region Campos

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data CampoInformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<CampoInformeDTO>
     */
    RespuestaPorDefectoAuditoria<List<CampoInformeDTO>> obtenerCamposPorIdPlantilla(HttpServletRequest httpServletRequest,
                                                                                    PlantillaInformeDTO plantillaInformeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data CampoInformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<CampoInformeDTO>
     */
    RespuestaPorDefectoAuditoria<List<CampoInformeDTO>> obtenerCamposPorIdInforme(HttpServletRequest httpServletRequest,
                                                                                  InformeDTO informeDTO);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data CampoInformeDTO
     *
     * @param httpServletRequest request peticion.
     * @return RespuestaPorDefectoAuditoria<CampoInformeDTO>
     */
    RespuestaPorDefectoAuditoria<List<CampoInformeDTO>> obtenerCamposPorNemonico(HttpServletRequest httpServletRequest,
                                                                                 String nemonico);
    //endregion
}
